src/wallet/spend.cpp

```diff
@@ -489,31 +489,19 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
     // Vector of results. We will choose the best one based on waste.
     std::vector<SelectionResult> results;
 
-    // Note that unlike KnapsackSolver, we do not include the fee for creating a change output as BnB will not create a change output.
-    std::vector<OutputGroup> positive_groups = GroupOutputs(wallet, available_coins, coin_selection_params, eligibility_filter, true /* positive_only */);
+    std::vector<OutputGroup> positive_groups = GroupOutputs(wallet, available_coins, coin_selection_params, eligibility_filter, /*positive_only=*/true);
     if (auto bnb_result{SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change)}) {
         results.push_back(*bnb_result);
     }
 
     // The knapsack solver has some legacy behavior where it will spend dust outputs. We retain this behavior, so don't filter for positive only here.
-    std::vector<OutputGroup> all_groups = GroupOutputs(wallet, available_coins, coin_selection_params, eligibility_filter, false /* positive_only */);
-    CAmount target_with_change = nTargetValue;
-    // While nTargetValue includes the transaction fees for non-input things, it does not include the fee for creating a change output.
-    // So we need to include that for KnapsackSolver and SRD as well, as we are expecting to create a change output.
-    if (!coin_selection_params.m_subtract_fee_outputs) {
-        target_with_change += coin_selection_params.m_change_fee;
-    }
-    if (auto knapsack_result{KnapsackSolver(all_groups, target_with_change, coin_selection_params.m_min_change_target, coin_selection_params.rng_fast)}) {
+    std::vector<OutputGroup> all_groups = GroupOutputs(wallet, available_coins, coin_selection_params, eligibility_filter, /*positive_only=*/false);
+    if (auto knapsack_result{KnapsackSolver(all_groups, nTargetValue, coin_selection_params.m_min_change_target, coin_selection_params.rng_fast)}) {
```

Comment:
Just a headscratcher: 
In the commit message of “wallet: accurate SelectionResult::m_target” (98d5dd26aec30ec63b555afc844281a4a04bd0fb) you write:

> SelectionResult::m_target should be equal to actual selection target.
> Selection target is the sum of all recipient amounts plus non input fees.
> So we need to remove change_fee from the m_target. It's safe because change
> target is always greater than the change fee, so we can always cover fees
> if change output is created.

It's not clear to me why the change target is always greater than change fee. Are you just alluding to the minimum change being more than the change fee even beyond 1,000 s/vB, or does the change fee actually get added anywhere that I'm missing?

Reply:
Good catch. Indeed what I meant is that `m_change_target` is at least `CHANGE_LOWER` which in most cases should be bigger than `change_fee`, but at high enough fee rate it doesn't hold.

Maybe we can ensure that `m_change_target` is at least `change_fee` by making `CHANGE_LOWER` dynamic and equal to max(50000, change_fee)`. @Xekyo @achow101 wdyt?

Reply:
@furszy #25825

Reply:
I've added c8cf08ea743e430c2bf3fe46439594257b0937e5 to address this. Please review